com.unity.shadergraph/Editor/Data/Nodes/NodeClassCache.cs

```diff
@@ -0,0 +1,112 @@
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Profiling;
+
+namespace UnityEditor.ShaderGraph
+{
+    [InitializeOnLoad]
+    public static class NodeClassCache
+    {
+        private static Dictionary<Type,List<ContextFilterableAttribute>> m_KnownTypeLookupTable;
+
+        private static Dictionary<Type, List<ContextFilterableAttribute>> KnownTypeLookupTable
+        {
+            get
+            {
+                if(m_KnownTypeLookupTable == null)
+                {
+                    ReCacheKnownNodeTypes();
+                }
+                return m_KnownTypeLookupTable;
+            }
+        }
+
+        public static IEnumerable<Type> KnownNodeTypes
+        {
+            get => KnownTypeLookupTable.Keys;
+        }
+
+        public static IEnumerable<ContextFilterableAttribute> GetFilterableAttributesOnNodeType(Type nodeType)
+        {
+            if (nodeType == null)
+            {
+                throw new ArgumentNullException("Cannot get attributes on a null Type");
+            }
+
+            if (KnownTypeLookupTable.TryGetValue(nodeType, out List<ContextFilterableAttribute> filterableAttributes))
+            {
+                return filterableAttributes;
+            }
+            else
+            {
+                throw new ArgumentException($"The passed in Type {nodeType.FullName} was not found in the loaded assemblies as a child class of AbstractMaterialNode");
+            }
+        }
+
+        public static T GetAttributeOnNodeType<T>(Type nodeType) where T : ContextFilterableAttribute
+        {
+            var filterableAttributes = GetFilterableAttributesOnNodeType(nodeType);
+            foreach(var attr in filterableAttributes)
+            {
+                if(attr is T searchTypeAttr)
+                {
+                    return searchTypeAttr;
+                }
+            }
+            return null;
+        }
+
+        private static void ReCacheKnownNodeTypes()
+        {
+            Profiler.BeginSample("NodeClassCache: Re-caching all known node types");
+            m_KnownTypeLookupTable = new Dictionary<Type, List<ContextFilterableAttribute>>();
+            foreach (Type nodeType in TypeCache.GetTypesDerivedFrom<AbstractMaterialNode>())
+            { 
+               if (nodeType.IsClass && !nodeType.IsAbstract)
```

Comment:
I don't think `IsClass` is needed when reading from `TypeCache.GetTypesDerivedFrom<AbstractMaterialNode>()`
```suggestion
               if (!nodeType.IsAbstract)
```